### PR TITLE
std: remove meta.trait

### PR DESCRIPTION
**Pull request by andrewrk**
In general, I don't like the idea of std.meta.trait, and so I am providing some guidance by deleting the entire namespace from the standard library and compiler codebase.

My main criticism is that it's overcomplicated machinery that bloats compile times and is ultimately unnecessary given the existence of Zig's strong type system and reference traces.

Users who want this can create a third party package that provides this functionality.

closes #18051

### CODE REVIEW

**Comment by nektro** [comment]
`std.meta.trait.isZigString` should be moved into the main `std.meta` namespace then, as this is an instrumentally useful function

**Review by dweiller — COMMENTED** [review]
I think there are a few places where old checks should be rewritten rather than removed.

In addition to those below, it might also be worth rewriting the checks in the `*As*ReturnType` functions in `mem.zig`, particularly for `asBytes`/`AsBytesReturnType` which don't do any field accesses that might help catch incorrect usage, but I'm less sure these cases.

**Comment by mitchellh** [comment]
> `std.meta.trait.isZigString` should be moved into the main `std.meta` namespace then, as this is an instrumentally useful function

What is this useful for? I don't mean that in a combative tone, this is a genuine question since I've never had a use for it.

File: lib/std/io/bit_writer.zig [hunk 1]
```diff
@@ -35,7 +33,6 @@ pub fn BitWriter(comptime endian: std.builtin.Endian, comptime WriterType: type)
             if (bits == 0) return;
 
             const U = @TypeOf(value);
-            comptime assert(trait.isUnsignedInt(U));
```
[the remark this answers is not in the document]
Comment (andrewrk) [reply lib/std/io/bit_writer.zig:38]: I see, thank you

**Comment by nektro** [comment]
when you're accepting a value in from an `anytype` and want to know if it is in the general shape of a `[]const u8`. `isZigString` handles checking for slice, nul-terminated pointer, pointer-to-array, etc

**Comment by xdBronch** [comment]
if youre allowing a function to accept both slices/array pointers and null terminated many pointers then you have to handle them differently regardless in which case using a trait function would be redundant, no?

**Comment by Jarred-Sumner** [comment]
looks like we have about 70 usages of `std.meta.trait.` in Bun's codebase

image

**Comment by andrewrk** [comment]
It's pretty straightforward to copy those trait functions into the bun codebase if you want to keep using them.

Looking at the screenshot, I would consider every one of those a code smell.

**Comment by andrewrk** [comment]
Buckle up, I have even more breakages planned.

**Comment by wrongnull** [comment]
> looks like we have about 70 usages of `std.meta.trait.` in Bun's codebase
> image

I decided to create a third party package that would retain this functionality and add new functionality depending on user needs in the future. https://github.com/wrongnull/zigtrait

**Pull request merged**